torchx/schedulers/ray_scheduler.py

```diff
@@ -0,0 +1,237 @@
+
+        # pyre-ignore[24]: Generic type `type` expects 1 type parameter
+        def set_job_attr(cfg_name: str, cfg_type: Type) -> None:
+            cfg_value = cfg.get(cfg_name)
+            if cfg_value is None:
+                return
+
+            if not isinstance(cfg_value, cfg_type):
+                raise TypeError(
+                    f"The configuration value '{cfg_name}' must be of type {cfg_type.__name__}."
+                )
+
+            setattr(job, cfg_name, cfg_value)
+
+        set_job_attr("cluster_name", str)
+        set_job_attr("start_cluster", bool)
+        set_job_attr("copy_scripts", bool)
+        set_job_attr("copy_script_dirs", bool)
+        set_job_attr("verbose", bool)
+
+        for role in app.roles:
+            # Replace the ${img_root}, ${app_id}, and ${replica_id} placeholders
+            # in arguments and environment variables.
+            role = macros.Values(
+                img_root=role.image, app_id=app_id, replica_id="${rank}"
+            ).apply(role)
+
+            actor = RayActor(
+                name=role.name,
+                command=" ".join([role.entrypoint] + role.args),
+                env=role.env,
+                num_replicas=max(1, role.num_replicas),
+                num_cpus=max(1, role.resource.cpu),
+                num_gpus=max(0, role.resource.gpu),
+            )
+
+            job.actors.append(actor)
+
+            if job.copy_scripts or job.copy_script_dirs:
+                # Find out the actual user script.
+                for arg in role.args:
+                    if arg.endswith(".py"):
+                        job.scripts.add(arg)
+
+        return AppDryRunInfo(job, repr)
+
+    def _validate(self, app: AppDef, scheduler: SchedulerBackend) -> None:
+        if scheduler != "ray":
+            raise ValueError(
+                f"An unknown scheduler backend '{scheduler}' has been passed to the Ray scheduler."
+            )
+
+        if app.metadata:
+            logger.warning("The Ray scheduler does not use metadata information.")
+
+        for role in app.roles:
+            if role.resource.capabilities:
+                logger.warning(
+                    "The Ray scheduler does not support custom resource capabilities."
+                )
+                break
+
+        for role in app.roles:
+            if role.port_map:
+                logger.warning("The Ray scheduler does not support port mapping.")
+                break
+
+    def _cancel_existing(self, app_id: str) -> None:
+        # TBD
+        pass
+
+    def describe(self, app_id: str) -> Optional[DescribeAppResponse]:
+        # TBD
+        return None
+
+    def log_iter(
+        self,
+        app_id: str,
+        role_name: str,
+        k: int = 0,
+        regex: Optional[str] = None,
+        since: Optional[datetime] = None,
+        until: Optional[datetime] = None,
+        should_tail: bool = False,
+    ) -> Iterable[str]:
+        # TBD
+        return []
```

Comment:
would be good to just leave these unimplemented / raise an Unimplemented error for now